web/src/message_list_view.js

```diff
@@ -619,101 +649,119 @@ export class MessageListView {
+                return; // No more groups to create
+            }
+            current_group = populate_group_from_message(
+                message_for_next_group,
+                same_day(message_for_next_group, prev_message),
+                this.get_possible_group_subscription_markers(prev_message, message_for_next_group),
+            );
         };
 
-        for (const message_container of message_containers) {
-            const message_reactions = reactions.get_message_reactions(message_container.msg);
-            message_container.msg.message_reactions = message_reactions;
-            message_container.include_recipient = false;
+        for (const message of messages) {
+            const message_reactions = reactions.get_message_reactions(message);
+            message.message_reactions = message_reactions;
+
+            // These will be used to build the message container
+            let include_recipient = false;
+            let subscribed;
+            let unsubscribed;
+            let stream_url;
+            let topic_url;
+            let pm_with_url;
+            let include_sender;
+            let want_date_divider;
+            let date_divider_html;
 
             if (
-                same_recipient(prev, message_container) &&
+                prev &&
+                util.same_recipient(prev.msg, message) &&
                 this.collapse_messages &&
-                prev.msg.historical === message_container.msg.historical
+                prev.msg.historical === message.historical
             ) {
-                add_message_container_to_group(message_container);
                 const date_divider_data = get_message_date_divider_data({
                     prev_message: prev.msg,
-                    curr_message: message_container.msg,
+                    curr_message: message,
                 });
-                message_container.want_date_divider = date_divider_data.want_date_divider;
-                message_container.date_divider_html = date_divider_data.date_divider_html;
+                want_date_divider = date_divider_data.want_date_divider;
+                date_divider_html = date_divider_data.date_divider_html;
             } else {
-                finish_group();
-                current_group = start_group();
-                add_message_container_to_group(message_container);
-
-                update_group_date(current_group, message_container.msg, prev?.msg);
-                message_container.want_date_divider = false;
-                message_container.date_divider_html = undefined;
-
-                message_container.include_recipient = true;
-                message_container.subscribed = false;
-                message_container.unsubscribed = false;
-
-                this.maybe_add_subscription_marker_to_group(
-                    current_group,
-                    prev?.msg,
-                    message_container.msg,
-                );
```

Comment:
reviewer note: This went into `populate_group_from_message` in `finish_group` -- maybe I should make a `start_group` to make this clearer? And a commit where I move this code there.